proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1208 @@
+
+By relying on borrowing, `Span` can provide simultaneous access to a non-copyable container, and can help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see the future directions for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+`Span` is the currency type for local processing over values in contiguous memory. It is the replacement for any API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that does not need to escape the value.
+
+### `RawSpan`
+
+`RawSpan` allows sharing the contiguous internal representation for values which may be heterogeneously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straightforward understanding of performance in library code.
+
+`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property or a constructor.
+
+## <a name="Design"></a>Detailed design
+
+`Span<Element>` is a simple representation of a span of initialized memory.
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  subscript(_ position: Int) -> Element { _read }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the `Span` from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
+The first element of a given span is _always_ at position zero, and its last it as position `count-1`.
+
+As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:
```

Comment:
Why does it not conform to `Collection` or `Sequence`? (I assume it's a non-copyable types issue, but it'd be good to explain it here.)

Reply:
This is explained in the introduction:
> Note that Span does not conform to Collection. This is because Collection, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a Collection is semantically a separate value from the instance it was derived from. In the case of Span, a sub-span representing a subrange of its elements must have the same lifetime as the Span from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of ~Copyable and ~Escapable for the collection and its elements.